### PR TITLE
Fixed issue #18655: Data loss if field in response table it missing

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Review by Shnoulle — COMMENTED** [review]
Unsure on Exception : i think we already have other : $ex->getMessage() ?

`sprintf($this->gT('The data could not be saved dur to %s'),$ex->getMessage())` maybe ?

File: application/helpers/expressions/em_manager_helper.php [hunk 1]
```diff
@@ -5159,7 +5159,18 @@ private function _UpdateValuesInDatabase($finished = false)
...
-                    $oResponse->setAttributes($aResponseAttributes, false);
+                    try {
+                        $oResponse->setAttributes($aResponseAttributes, false);
+                    } catch (Exception $ex) {
+                        // This can happen if the table is missing fields. It should never happen, but somehow it does.
+                        submitfailed($ex->getMessage());
+                        if (YII_DEBUG) {
+                            throw $ex;
+                        }
+                        $message = $this->gT('The data could not be saved because the database doesn\'t match the survey structure.');
```
[the remark this answers is not in the document]
Comment (c-schmitz) [reply application/helpers/expressions/em_manager_helper.php:5170]: Please change the message to:
```"Due to a technical problem, your response could not be saved. Please contact the survey administrator <Survey administrator name> <Survey administrator email address> about this problem. You will not be able to proceed with this survey."```

I also think that at this point the survey session should be killed.

**Review by c-schmitz — CHANGES_REQUESTED** [review]

File: application/helpers/expressions/em_manager_helper.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/helpers/expressions/em_manager_helper.php:5170]: > Unsure it this point, no ? We already have some other error (for example id of response deleted during survey)

We are kind of using the same mechanism used couple of lines below.
https://github.com/LimeSurvey/LimeSurvey/pull/2946/files#diff-fa8d35264aa59c85d37ca2d1756472996499f06fb346865594378102f5f7abc0L5165

**Comment by gabrieljenik** [comment]
> `sprintf($this->gT('The data could not be saved dur to %s'),$ex->getMessage())` maybe ?

We didn't want to show technical info to the participant.
So we send one message to the admin and show a different one on the front end.

**Review by c-schmitz — APPROVED** [review]

**Comment by c-schmitz** [comment]
@gabrieljenik Had to revert this. This would break response export if a token table exists.

**Comment by gabrieljenik** [comment]
Sorry about this. I will make the nwe method to have a different name, so we can control from where it is being used.

**Pull request merged**